### PR TITLE
Permissions for Polls

### PR DESCRIPTION
**Pull request by Bartzi**
This PR enables us to provide permissions for polls, thereby refactoring the Permission form to be more general and easy to adapt to new situations.

The code has two pitfalls:

1. I was not able to come up with a nicer rendering of permission names for each group. Does s.b. have a better idea?
2. I had to replace the `as_table ` method of the form. This is because I did not see another way of rendering the form in a generic way and the standard as_table method does not render the html the way we want. If somebody has any suggestions...

Please also make sure to review whether I checked that all permissions are checked at the right place at the right time.

I will add tests later on. That's why this PR is WIP

### CODE REVIEW

File: _1327/polls/templates/polls_create_poll.html [hunk 1]
```diff
@@ -39,6 +39,19 @@
...
+        <h3>{% trans "Permissions" %}</h3>
+        <div class="divider"></div>
+        {{ permission_formset.management_form }}
+        <table class="table table-striped vertically-aligned table-narrow" >
+			<thead>
+				{{ permission_header }}
+			</thead>
+			<tbody>
+			{% for form_element in permission_formset %}
+                {{ form_element.as_table }}
```
Comment (Nef10) [inline _1327/polls/templates/polls_create_poll.html]: use tabs

File: _1327/documents/forms.py [hunk 2]
```diff
@@ -40,39 +41,70 @@ def clean_url_title(self):
...
+			else:
+				remove_perm(field_name, group, model)
+
+	@classmethod
+	def header(cls, content_type):
+		output = [
+			'<tr>',
+			'<th class="col-md-6"> {} </th>'.format(_("Role")),
+		]
+		for permission in sorted(Permission.objects.filter(content_type=content_type), key=lambda x: x.codename):
```
Comment (invliD) [inline _1327/documents/forms.py]: Try using the fields instead of a separate database query.

File: _1327/documents/forms.py [hunk 3]
```diff
@@ -40,39 +42,70 @@ def clean_url_title(self):
...
+			if value:
+				assign_perm(field_name, group, model)
+			else:
+				remove_perm(field_name, group, model)
+
+	@classmethod
+	def header(cls):
+		output = [
+			'<tr>',
+			'<th class="col-md-6"> {} </th>'.format(_("Role")),
```
Comment (invliD) [inline _1327/documents/forms.py]: Remove spaces around `{}`

File: _1327/documents/forms.py [hunk 4]
```diff
@@ -40,39 +42,70 @@ def clean_url_title(self):
...
+		]
+		for permission in sorted([name for name, field in filter(lambda x: type(x[1]) == BooleanField, cls.base_fields.items())]):
+			item = "<th class=\"col-md-2 text-center\"> {} </th>".format(_(permission.rsplit('_')[0]))
```
Comment (invliD) [inline _1327/documents/forms.py]: Same here.

File: _1327/documents/forms.py [hunk 5]
```diff
@@ -40,39 +42,70 @@ def clean_url_title(self):
...
+			else:
+				remove_perm(field_name, group, model)
+
+	@classmethod
+	def header(cls):
+		output = [
+			'<tr>',
+			'<th class="col-md-6"> {} </th>'.format(_("Role")),
+		]
+		for permission in sorted([name for name, field in filter(lambda x: type(x[1]) == BooleanField, cls.base_fields.items())]):
```
Comment (invliD) [inline _1327/documents/forms.py]: Though I could figure out what this does, please add a comment.

File: _1327/documents/forms.py [hunk 6]
```diff
@@ -40,39 +42,70 @@ def clean_url_title(self):
...
+			item = "<th class=\"col-md-2 text-center\"> {} </th>".format(_(permission.rsplit('_')[0]))
+			output.append(item)
+		output.append('</tr>')
+		return mark_safe('\n'.join(output))
```
Comment (invliD) [inline _1327/documents/forms.py]: If you mark this as safe, you should escape all the stuff going in the HTML.

File: _1327/documents/forms.py [hunk 7]
```diff
@@ -40,39 +42,70 @@ def clean_url_title(self):
...
+		]
+
+		for name in sorted(self.fields.keys()):
+			if name == "group_name":
+				continue
+			output.append('<td class="text-center"> {} </td>'.format(self[name]))
+		output.append(str(self['group_name']))
+		output.append("</tr>")
+
+		return mark_safe('\n'.join(output))
```
Comment (invliD) [inline _1327/documents/forms.py]: Same here.

File: _1327/documents/forms.py [hunk 8]
```diff
@@ -40,39 +42,70 @@ def clean_url_title(self):
...
+		"Returns this form rendered as HTML <tr>"
+		output = [
+			"<tr>",
+			"<td>{}</td>".format(self['group_name'].value())
+		]
+
+		for name in sorted(self.fields.keys()):
+			if name == "group_name":
+				continue
+			output.append('<td class="text-center"> {} </td>'.format(self[name]))
```
Comment (invliD) [inline _1327/documents/forms.py]: Same here.

**Comment by janno42** [comment]
the `add` permission should not be shown in the table

**Comment by janno42** [comment]
when any permission for a group gets selected, the clean method should check that the `view` permission is also selected.

**Comment by janno42** [comment]
the `permission_warning` (documents/utils.py) is defined for InformationDocuments, so when used for minutes the warning is always shown.

**Pull request closed without merging**